src/NuGet.Core/NuGet.Protocol/Utility/LocalFolderUtility.cs

```diff
@@ -1202,5 +1279,53 @@ public static void GenerateNupkgMetadataFile(string nupkgPath, string installPat
                     }
                 });
         }
+
+        private static T[] CancellableEnumaration<T>(IEnumerable<T> enumerable, ILogger log, CancellationToken cancellationToken)
+        {
+            try
+            {
+                // .ToArray is necessary because previously it returns Array
+                // If enumaration happen several times then same I/O calls repeatedly called on same input, I/O calls are more expensive then memory.
+                return CancellableYieldEnumaration(enumerable, log, cancellationToken).ToArray();
+            }
+            catch (Exception e) when (e is not OperationCanceledException)
+            {
+                // On cancellation we bubble up exception to call stack.
+                // Otherwise return all or nothing. If no exception return all.
+                // Break on first exception with logging in order to keep previous experience, return empty array, but don't throw.
+                log.LogWarning(e.Message);
+            }
+
+            return Array.Empty<T>();
+        }
+
+        private static IEnumerable<T> CancellableYieldEnumaration<T>(IEnumerable<T> enumerable, ILogger log, CancellationToken cancellationToken)
+        {
+            using (var enumerator = enumerable.GetEnumerator())
```

Comment:
Now I'm a little confused, why isn't this just a foreach at this point?

I just want to make sure I call out the behavior;

- Prior behavior was that if any I/O exception occurred, no results would change
- Next, you implemented it if that any I/O exception occurred, all the results up until that would be returned
- Now you are back to the former behavior, which is okay - but then there's no need to manually walk the IEnumerator just foreach.

Reply:
Sure, `foreach` make sense now.  It looks I started over engineering this.

Reply:
Converted to draft. Unit test is failing.

Reply:
Fixed failing test. Had to remove `CancellableEnumeration<T>`, try catch need to cover  `root.EnumerateDirectories(filter, searchOption);` to detect long filter name. I may use `func` to keep `CancellableEnumeration<T>`, but want to avoid additional alloc due to func.